src/index.js

```diff
@@ -204,381 +253,378 @@ class Masq {
     return JSON.parse(currentUserInfo).profileImage
   }
 
-  async _initSwarmWithDataHandler (channel, dataHandler) {
-    return new Promise((resolve, reject) => {
+  async _genLoginLink () {
+    this.loginChannel = uuidv4()
+    this.loginKey = await common.crypto.genAESKey(true, 'AES-GCM', 128)
+    const extractedKey = await common.crypto.exportKey(this.loginKey)
+    const keyBase64 = Buffer.from(extractedKey).toString('base64')
+    this.loginUrl = new URL(this.config.masqAppBaseUrl)
+    const requestType = 'login'
+    const hashParams = JSON.stringify([this.appName, requestType, this.loginChannel, keyBase64])
+    this.loginUrl.hash = '/link/' + Buffer.from(hashParams).toString('base64')
+
+    this.loginLink = this.loginUrl.href
+  }
+
+  async getLoginLink () {
+    return this.loginLink
+  }
+
+  async logIntoMasq (stayConnected) {
+    this.setState('loggingIn')
+
+    // make stayConnected i boolean
+    this.stayConnected = !!stayConnected
+
+    let pr = new Promise((resolve, reject) => {
       // Subscribe to channel for a limited time to sync with masq
-      debug(`Creation of a hub with ${channel} channel name`)
+      debug(`Creation of a hub with ${this.loginChannel} channel name`)
 
-      const hub = signalhub(channel, this.config.hubUrls)
+      const hub = signalhub(this.loginChannel, this.config.hubUrls)
       hub.on('error', ({ url, error }) => {
         reject(new MasqError(MasqError.SIGNALLING_SERVER_ERROR, url, error))
       })
-      const sw = this._createSwarm(hub)
 
-      sw.on('peer', (peer, id) => {
+      this.loginSw = this._createSwarm(hub)
+
+      let dataReceived = false
+
+      this.loginSw.on('peer', (peer, id) => {
```

Comment:
We have to fix this code to allow the cancellation of the callback in case while executing it a disconnect happens. We can reformat this code when we fix this bug.